solr/core/src/java/org/apache/solr/search/facet/FacetRequestSortedMerger.java

```diff
@@ -232,7 +232,9 @@ public Map<String, Object> getRefinement(Context mcontext) {
       return null;
     }
 
-    final FacetRequest.FacetSort initial_sort = null == freq.prelim_sort ? freq.sort : freq.prelim_sort;
+    // Use final sort here (as opposed to prelim_sort if available) to determine top buckets for refinement.
+    // At this point, we want to take our best guess at the final sorted values for refinement.
+    final FacetRequest.FacetSort sort = freq.sort;
```

Comment:
Maybe this should be an option if/when upstreamed? But I truly feel that this is _far_ better here. Refinement is already operating on results that were prelim sorted, truncated, then resorted on a per shard basis. It should pick which buckets to refine based on estimated final sorting when we are working at the merge level. Thoughts?

Reply:
I _think_ this makes sense?

Trying to play devil's advocate though: I wonder if it depends on the particular kind of sort? Practically speaking, most cases would be prelim_sort by `count desc`. I think `prelim_sort` was initially introduced for sort-by-relatedness (correlation); a little hand-wavy here, but I think `processEmpty` can be important for correlation, filling in background set counts that affect relatedness calculation. It's possible that prematurely calculating `relatedness` based on incomplete information could skew results _more_ than determining refinement based on prelim sort-by-count -- particularly for the case where values might be correlated to particular shards (dependent on shard routing scheme)?

Reply:
Yeah, I think that makes sense as well. I'm just going to throw in an option for this one :).